### DISCUSSION TITLE
using ivabs cross-section optimization

### DISCUSSION THREAD

**Discussion by kimjihun8** [wenbinyugroup/.github#18] [VABS Related Discussions] [answered]
I want to use ivabs to perform optimization to move the center of gravity of the section to a specific location by adjusting the location of the spar web or the thickness of the laminate of the section.

The constraint at this time is to ensure that the strain is below a specific value in the structural analysis using the beam with the section characteristics applied.
The strain will be the axial strain of the beam as shown in the figure below.
![strain](https://github.com/user-attachments/assets/b2f5e79f-587f-4c21-81e2-17dec066bd8b)

What I am wondering is whether it is possible to optimize using the strain as a constraint when using ivabs at this time.
Since I have little experience using dakota and gebt included in ivabs, it is difficult to know if this is possible.
Please advise.

**Comment by wenbinyu-PU** [comment]
I am referring to @unitasium to answer this question.

**Comment by unitasium** [comment]
Right now using strains as constraints is not available yet. We can add this feature to the next update. Specifically, how would you like to use the strain data? Do you plan to use VABS built-in failure criterion as the constraint or use the local 3D strain field with custom post-processing script?

**Comment by unitasium** [comment]
Could you clarify the following questions:
1. Which strain are you referring to in the constraint, the global beam strain or the local 3D solid strain?
2. Is the constraint function a simple comparison between values (check if global beam strain < specified value) or a failure criterion-based constraint which requires recovering of local 3D fields?

**Reply by kimjihun8** [reply]
1. As shown in the figure, it is a global beam strain.
2. Similar to 'check if global beam strain < specified value'. I would check in the form of 'less than a certain % of specified value'.

**Reply by unitasium** [reply, marked as answer]
You can use a `script` type analysis step to call a python function to 1) extract data from beam analysis output, 2) process the data to calculate the quantity used in the constraint. Then you can specify the upper bound of the constraint in the `study` block of the ivabs main input.

**Comment by kimjihun8** [comment]
1. As shown in the figure, it is a global beam strain.
2. Similar to 'check if global beam strain < specified value'. I would check in the form of 'less than a certain % of specified value'.
